cmd/hack/hack.go

```diff
@@ -1666,6 +1681,243 @@ func GenerateTxLookups1(chaindata string, block int) {
+	copy(a[35*i:35*i+35], a[35*j:35*j+35])
+	copy(a[35*j:35*j+35], tmp)
+}
+
+func insertInFileForLookups2(file *os.File, entries Entries, it uint64) {
+	sorted := entries[:35*it]
+	sort.Sort(sorted)
+	_, err := file.Write(sorted)
+	check(err)
+	log.Info("File Insertion Occured")
+}
+
+func generateTxLookups2(db *ethdb.BoltDatabase, startBlock uint64, interruptCh chan bool) {
+	var bufferLen int = 143360 // 35 * 4096
+	var count uint64 = 5000000
+	var entries Entries = make([]byte, count*35)
+	bn := make([]byte, 3)
+	var lookups []LookupFile
+	var iterations uint64
+	var blockNum = startBlock
+	var interrupt bool
+	filename := fmt.Sprintf(".lookups_%d.tmp", len(lookups))
+	fileTmp, _ := os.OpenFile(filename, os.O_APPEND|os.O_CREATE|os.O_RDWR, 0644)
+	for !interrupt {
+		blockHash := rawdb.ReadCanonicalHash(db, blockNum)
+		body := rawdb.ReadBody(db, blockHash, blockNum)
+
+		if body == nil {
+			log.Info("Now Inserting to file")
+			insertInFileForLookups2(fileTmp, entries, iterations)
+			lookups = append(lookups, LookupFile{nil, nil, 35})
+			iterations = 0
+			entries = nil
+			fileTmp.Close()
+			break
+		}
+
+		select {
+		case interrupt = <-interruptCh:
+			log.Info("interrupted, please wait for cleanup...")
+		default:
+		}
+		bn[0] = byte(blockNum >> 16)
+		bn[1] = byte(blockNum >> 8)
+		bn[2] = byte(blockNum)
+
+		for _, tx := range body.Transactions {
+			copy(entries[35*iterations:], tx.Hash().Bytes())
+			copy(entries[35*iterations+32:], bn)
+			iterations++
+			if iterations == count {
+				log.Info("Now Inserting to file")
+				insertInFileForLookups2(fileTmp, entries, iterations)
+				entries = nil
+				lookups = append(lookups, LookupFile{nil, nil, 35})
+				iterations = 0
+				fileTmp.Close()
+				filename = fmt.Sprintf(".lookups_%d.tmp", len(lookups))
+				fileTmp, _ = os.OpenFile(filename, os.O_APPEND|os.O_CREATE|os.O_RDWR, 0644)
+			}
+		}
+		blockNum++
+		if blockNum%100000 == 0 {
+			log.Info("Processed", "blocks", blockNum, "iterations", iterations)
+			var m runtime.MemStats
+			runtime.ReadMemStats(&m)
+			log.Info("Memory", "alloc", int(m.Alloc/1024), "sys", int(m.Sys/1024), "numGC", int(m.NumGC))
+		}
+	}
+	batch := db.NewBatch()
+	h := &Heap{}
+	heap.Init(h)
+	for i := range lookups {
+		file, err := os.Open(fmt.Sprintf(".lookups_%d.tmp", i))
+		check(err)
+		lookups[i].file = bufio.NewReader(file)
+		lookups[i].buffer = make([]byte, bufferLen)
+		check(err)
+		n, err := lookups[i].file.Read(lookups[i].buffer)
+		if n != bufferLen {
+			lookups[i].buffer = lookups[i].buffer[:n]
+		}
+		heap.Push(h, HeapElem{lookups[i].buffer[:35], i})
+	}
+
+	for !interrupt && len(*h) != 0 {
+		val := (heap.Pop(h)).(HeapElem)
+		if lookups[val.index].pos == uint64(bufferLen) {
+			if val.val[32] != 0 {
+				err := batch.Put(dbutils.TxLookupPrefix, val.val[:32], common.CopyBytes(val.val[32:]))
+				check(err)
+			} else {
+				err := batch.Put(dbutils.TxLookupPrefix, val.val[:32], common.CopyBytes(val.val[33:]))
```

Comment:
Why is it `val.val[33:]` here and `val.val[32:]` up there? Is this something to do with the encoding of block numbers?

Reply:
yes, in case of trailing 0s